lib/dap/filter/geoip2.rb

```diff
@@ -0,0 +1,325 @@
+    end
+    return unless (geo_hash = @@geo_city.get(ip))
+    ret = defaults
+
+    if geo_hash.include?("subdivisions")
+      # handle countries that are divided into various subdivisions.  generally 1, sometimes 2
+      subdivisions = geo_hash["subdivisions"]
+      geo_hash.delete("subdivisions")
+      ret["geoip2.city.subdivisions.length"] = subdivisions.size.to_s
+      subdivisions.each_index do |i|
+        subdivision = subdivisions[i]
+        subdivision.each_pair do |k,v|
+          if %w(geoname_id iso_code).include?(k)
+            ret["geoip2.city.subdivisions.#{i}.#{k}"] = v.to_s
+          elsif k == "names"
+            if v.include?(GEOIP2_LANGUAGE)
+              ret["geoip2.city.subdivisions.#{i}.name"] = subdivision["names"][GEOIP2_LANGUAGE]
+            end
+          end
+        end
+      end
+    end
+
+    Dap::Utils::Misc.flatten_hash(geo_hash).each_pair do |k,v|
+      if DESIRED_GEOIP2_KEYS.include?(k)
+        # these keys we can just copy directly over
+        ret["geoip2.city.#{k}"] = v
+      elsif @locale_specific_names.include?(k)
+        # these keys we need to pick the locale-specific name and set the key accordingly
+        lsn_renamed = k.gsub(/\.names.#{GEOIP2_LANGUAGE}/, ".name")
+        ret["geoip2.city.#{lsn_renamed}"] = v
+      end
+    end
+    ret
+  end
+
+  def defaults()
+    ret = {}
+    default_int_suffixes = %w(geoname_id metro_code)
+    default_bool_suffixes = %w(is_in_european_union is_anonymous_proxy is_satellite_provider)
+    DESIRED_GEOIP2_KEYS.each do |k|
+      suffix = k.split(/\./)[-1]
+      if default_int_suffixes.include?(suffix)
+        ret["geoip2.city.#{k}"] = "0"
+      elsif default_bool_suffixes.include?(suffix)
+        ret["geoip2.city.#{k}"] = "false"
+      else
+        ret["geoip2.city.#{k}"] = ""
+      end
+    end
+    ret
+  end
+end
+
+#
+# Add GeoIP2 ASN and Org tags using the MaxMind GeoIP2::ASN database
+#
+class FilterGeoIP2Asn
+  include BaseDecoder
+  include GeoIP2Library
+
+  def decode(ip)
+    unless @@geo_asn
+      raise "No MaxMind GeoIP2::ASN data found"
+    end
+    geo_hash = @@geo_asn.get(ip)
+    return unless geo_hash
+
+    ret = {}
+
+    if geo_hash.include?("autonomous_system_number")
+      ret["geoip2.asn.asn"] = "AS#{geo_hash["autonomous_system_number"]}"
+    else
+      ret["geoip2.asn.asn"] = ""
+    end
+
+    if geo_hash.include?("autonomous_system_organization")
+      ret["geoip2.asn.asn_org"] = "#{geo_hash["autonomous_system_organization"]}"
+    else
+      ret["geoip2.asn.asn_org"] = ""
+    end
+
+    ret
+  end
+end
+
+#
+# Add GeoIP2 ISP tags using the MaxMind GeoIP2::ISP database
+#
+class FilterGeoIP2Isp
+  include BaseDecoder
+  include GeoIP2Library
+  def decode(ip)
+    unless @@geo_isp
+      raise "No MaxMind GeoIP2::ISP data found"
+    end
+    geo_hash = @@geo_isp.get(ip)
+    return unless geo_hash
+
+    ret = {}
+
+    if geo_hash.include?("autonomous_system_number")
```

Comment:
this and all the conditionals below could probably be extracted into a function that modifies ret.  could DRY up the code a bit